README.md

```diff
@@ -1,2 +1,310 @@
-# validator
-A tiny runtime library for type assertions
+# Validator
+
+A tiny library for run time type assertions.
+
+## Installation
+
+```bash
+npm install @mapbox/validator
+```
+
+## Usage
+
+### Basic
+The basic usage 
+
+```javascript
+const v = require('@mapbox/validator');
+
+assertObj = v.assert(
+  v.shape({
+      names: v.arrayOf(v.string)
+  })
+);
+
+assertObj({names: ['ram', 'harry']}); // pass
+assertObj({names: ['john', 987]})// fail
+assertObj({names: 'john'})// fail
+```
+
+### Required
+
+Validation of each value is optional by default, which means passing `null` or `undefined` would not fail the assertion. You should wrap your validator with `v.required` in case you want strict assertion.
+
+```javascript
+assertObj = v.assert(
+  v.shape({
+      name: v.string
+  })
+);
+assertObj({}); // pass
+assertObj({name: undefined}); // pass
+assertObj({name: null}); // pass
+assertObj({name: 9}); // fail
+
+strictAssertObj = v.assert(
+     v.shape({
+      name: v.required(v.string)
+  })
+)
+strictAssertObj({}); // fail
+strictAssertObj({name: undefined}); // fail
+strictAssertObj({name: null}); // fail
+```
+
+### Composition
+You can compose any of the [Higher Order Validators](#higher-order-validators) to make complex validators.
+<details><summary>Example 1</summary>
```

Comment:
I'm not sure the `<details>` is necessary here. Maybe with a table of contents the page will seem less intimidating? I imagine that with a long Markdown file the main usage will be with the table of contents or ctrl+F, so most users would rather have the examples automatically visible instead of requiring a click.

That's another perspective to consider. Your call, based on what you think makes the most sense 👍

Reply:
My earlier idea was to reduce the length and make it easy to navigate. But ya I agree clicking adds additional barrier to consuming content. I have removed the click and added a toc.